swift/AssignNilToImplicitlyUnwrappedOptional.yaml

```diff
@@ -0,0 +1,33 @@
+rules:
+  - id: assign-nil-to-implicitly-unwrapped-optional
+    message: >
+      An implicitly unwrapped optional was assigned the value nil. If the
+      variable is dereferenced before it is assigned a non-nil value, a fatal
+      error will occur.
+    languages: [swift]
+    severity: WARNING
+    metadata:
+      category: security
+      subcategory:
+        - audit
+      technology:
+        - shell
+      cwe: "CWE-250: Execution with Unnecessary Privileges"
+      confidence: HIGH
+      likelihood: HIGH
+      impact: HIGH
+      references:
+        - https://docs.swift.org/swift-book/documentation/the-swift-programming-language/thebasics/#Implicitly-Unwrapped-Optionals
+      description: Implicitly unwrapped optionals are always unwrapped whenever they are dereferenced. Such a variable should never be assigned the value nil after initialization.
+    patterns:
+      - pattern-inside:
+          pattern-either:
+            - pattern: |
+                var $VAR: Int! = nil
+                (...)
+            - pattern: |
+                var $VAR: Int!
+                (...)
```

Comment:
Huh, I'm kinda surprised this runs. AFAIK, `pattern-inside` must take a string, not a child YAML object. IMO a better way to structure this would be to have the `pattern-either` as the parent pattern, and have two `pattern-inside`s underneath.

I think [taint](https://semgrep.dev/docs/writing-rules/data-flow/taint-mode) mode is another great option here. It perfectly matches what you're trying to achieve here. The `pattern`s above would be the sources, the `pattern-not`s would be sanitizers, and the `pattern` below would be the sink.

Reply:
Thanks for the review. Given CodeQL's inability to richly express variable types, I'm going to chuck this for a CodeQL version at some point in the future. Thanks!